### PR TITLE
Add LSTM text generator example and livebook

### PR DESCRIPTION
**Pull request by alisinabh**
This example is based on [this post](https://machinelearningmastery.com/text-generation-lstm-recurrent-neural-networks-python-keras/) by Jason Brownlee.

### CODE REVIEW

File: examples/generative/text_generator.exs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      init_seq
+      |> String.downcase()
+      |> String.to_charlist()
+
+    Enum.reduce(1..80, init_seq, fn _, seq ->
+      init_seq =
+        seq
+        |> Enum.reverse()
+        |> Enum.take(@sequence_length)
+        |> Enum.reverse()
```
Comment (seanmor5) [inline examples/generative/text_generator.exs:49]: I am confused why you need the double reverse?

**Review by seanmor5 — APPROVED** [review]
Wow, awesome work! Can you by chance also contribute this as a Livebook with step-by-step explanations and saved outputs? That way we can include it in the docs. It's okay if it's in a separate PR

File: examples/generative/text_generator.exs (hunk 1 above)
Comment (jonatanklosko) [inline examples/generative/text_generator.exs:49]:
```suggestion
        |> Enum.take(-@sequence_length)
```

File: examples/generative/text_generator.exs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        |> Enum.map(&Map.fetch!(char_to_int, &1))
+        |> Nx.tensor()
+        |> Nx.divide(Enum.count(characters))
+        |> Nx.reshape({1, @sequence_length, 1})
+
+      char =
+        Axon.predict(model, params, init_seq)
+        |> Nx.argmax()
+        |> Nx.to_number()
+        |> then(&Enum.at(characters, &1))
```
Comment (jonatanklosko) [inline examples/generative/text_generator.exs:59]: We are converting the number to char, but in the next iteration we convert all characters to numbers. One alternative would be to convert `init_seq` to numbers, work on numbers, and the map back to characters at the end?

File: examples/generative/text_generator.exs (hunk 1 above)
Comment (jonatanklosko) [reply examples/generative/text_generator.exs:49]: Is this because of the character we add at the end of each iteration? If so, then we can shift it there, like `tl(seq) ++ [char]`.

Comment (alisinabh) [reply examples/generative/text_generator.exs:49]: The reason is I need the last `@sequence_length` tokens for the next input but in correct order. The better approach as mentioned by @jonatanklosko is to use `Enum.take(-@sequence_length)`.

Comment (alisinabh) [reply examples/generative/text_generator.exs:49]: The first suggestion is better because in the second one we will loose some parts of the sentence especially if we generate more than sequence length.

Comment (alisinabh) [reply examples/generative/text_generator.exs]: Addressed in 0e6b697

File: examples/generative/text_generator.exs (hunk 2 above)
Comment (alisinabh) [reply examples/generative/text_generator.exs]: Good point. Addressed in 0e6b697

**Comment by alisinabh** [comment]
@seanmor5 Sure. I have actually made the example out of the livebook that I have created :sweat_smile: I've included it in this PR. Please let me know if there are any mistakes or confusion in it.

File: notebooks/text_generator.livemd [hunk 3]
```diff
@@ -0,0 +1,920 @@
...
+filename = "book.txt"
+# change download_url if you like to experiment with other text books.
+download_url = "https://www.gutenberg.org/files/11/11-0.txt"
+
+unless File.exists?(filename) do
+  Req.get!(download_url)
+  |> then(fn %{body: content} -> File.write!(filename, content) end)
+end
+
+book_text = File.read!(filename)
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:46]: Since the payload is not really big, we don't need to cache. Also, with this approach changing the URL wouldn't have effect until the file is deleted :)

```suggestion
# Change the URL if you like to experiment with other text books
download_url = "https://www.gutenberg.org/files/11/11-0.txt"

book_text = Req.get!(download_url).body
```

File: notebooks/text_generator.livemd [hunk 4]
```diff
@@ -0,0 +1,920 @@
...
+```
+
+First of all, we need to normalize the content of the book. We are only interested in the sequence of English characters, periods and new lines. Also currently we don't care about the capitalization and things like apostrophe so we can remove all other unknown characters and downcase everything. We can use regex for that.
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:55]:
```suggestion
First of all, we need to normalize the content of the book. We are only interested in the sequence of English characters, periods and new lines. Also currently we don't care about the capitalization and things like apostrophe so we can remove all other unknown characters and downcase everything. We can use a regular expression for that.
```

File: notebooks/text_generator.livemd [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+
+We can also convert the string into a list of characters so we can handle them easier. You will understand exactly why a bit further.
+
+```elixir
+normalized_book_text =
+  book_text
+  |> String.downcase()
+  |> String.replace(~r/[^A-Za-z \.\n]/, "")
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:63]: Since we downcase beforehand

```suggestion
  |> String.replace(~r/[^a-z \.\n]/, "")
```

File: notebooks/text_generator.livemd [hunk 6]
```diff
@@ -0,0 +1,920 @@
...
+```
+
+Lowercase English characters in ASCII encoding are presented with numbers between `97 = a` and `122 = z` a space is `32 = [ ]` a new line is `10 = \n` and the period is `46 = .`.
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:73]: More precisely, the charlist includes unicode code points, it's just that in our case it matches ASCII.

File: notebooks/text_generator.livemd [hunk 7]
```diff
@@ -0,0 +1,920 @@
...
+
+```elixir
+sequence_length = 100
+
+train_data =
+  normalized_book_text
+  |> Enum.map(&Map.fetch!(char_to_int, &1))
+  |> Enum.chunk_every(sequence_length, 1, :discard)
+  # We don't want the last chunk since we don't have a prediction for it.
+  |> Enum.slice(0..-2)
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:185]:
```suggestion
  |> Enum.drop(-1)
```

File: notebooks/text_generator.livemd [hunk 8]
```diff
@@ -0,0 +1,920 @@
...
+  [0, 0, 1]
+]
+```
+
+In Nx, you can achieve this encoding by using this snippet
+
+<!-- livebook:{"force_markdown":true} -->
+
+```elixir
+Nx.equal(tensor, Nx.tensor(Enum.to_list(0..last_feature)))
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:281]:
```suggestion
Nx.equal(tensor, Nx.iota({1, last_feature}))
```

Maybe make it an anoonymous function, and show how it works instead of the iex snippet above?

File: notebooks/text_generator.livemd [hunk 9]
```diff
@@ -0,0 +1,920 @@
...
+```elixir
+# As the input, We expect the sequence_length characters. Each one is a number.
+model =
+  Axon.input({nil, sequence_length, 1}, "input_chars")
+  # The Long Short Term Memory layer of our network
+  |> Axon.lstm(256)
+  # Selecting only the output from the LSTM Layer
+  |> then(fn {_, out} -> out end)
+  # Since we only want the last sequence in LSTM we will slice it and select the last one
+  |> Axon.nx(fn x -> Nx.slice_along_axis(x, sequence_length - 1, 1, axis: -2) end)
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:337]: I think if we do this, we no longer need flatten

```suggestion
  |> Axon.nx(fn x -> x[[0..-1//1, -1, 0..-1//1]] end)
```

File: notebooks/text_generator.livemd [hunk 10]
```diff
@@ -0,0 +1,920 @@
...
+train_batches = Nx.to_batched_list(train_data, batch_size)
+result_batches = Nx.to_batched_list(train_results, batch_size)
+
+IO.puts("Total batches: #{Enum.count(train_batches)}")
+
+params =
+  model
+  |> Axon.Loop.trainer(:categorical_cross_entropy, Axon.Optimizers.adam(0.001))
+  |> Axon.Loop.run(Stream.zip(train_batches, result_batches), %{}, epochs: 20, compiler: EXLA)
+```
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:392]: Unless you think showing the parameters makes sense, we can do this:

````suggestion

:ok
```
````

File: notebooks/text_generator.livemd [hunk 11]
```diff
@@ -0,0 +1,920 @@
...
+       ]
+     >}
+  }
+}
+```
+
+## Generating Text
+
+After last step we will have our trained network parameters so we can start generating text with it. We simply need to give the initial sequence as the input to the network and select the most confident output. `Axon.predict/3` will give us the output layer and then using `Nx.argmax/1` we get the most confident neuron index then simply convert that index back to it's ASCII presentation.
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:527]:
```suggestion
After last step we will have our trained network parameters so we can start generating text with it. We simply need to give the initial sequence as the input to the network and select the most confident output. `Axon.predict/3` will give us the output layer and then using `Nx.argmax/1` we get the most confident neuron index then simply convert that index back to it's ASCII representation.
```

(though maybe we should say Unicode code point instead)

File: notebooks/text_generator.livemd [hunk 12]
```diff
@@ -0,0 +1,920 @@
...
+      |> Nx.reshape({1, sequence_length, 1})
+
+    char =
+      Axon.predict(model, params, init_seq)
+      |> Nx.argmax()
+      |> Nx.to_number()
+
+    seq ++ [char]
+  end)
+  |> Enum.map(&Enum.at(characters, &1))
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:556]: We can create `int_to_char` and use that for more efficient mapping.

File: examples/generative/text_generator.exs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    normalized_book_text =
+      book_text()
+      |> String.downcase()
+      |> String.replace(~r/[^A-Za-z \.\n]/, "")
+      |> String.to_charlist()
+
+    # Extract all then unique characters we have. Optionally we can sort them.
+    characters = normalized_book_text |> Enum.uniq() |> Enum.sort()
+    # Create a mapping for every character
+    char_to_int = Enum.with_index(characters) |> Enum.into(%{})
```
Comment (jonatanklosko) [inline examples/generative/text_generator.exs:97]: It may be confusing that characters are already integers, perhaps if we name it `char_to_idx` it will be more clear?

**Review by jonatanklosko — COMMENTED** [review]
Awesome, a couple more comments :)

File: notebooks/text_generator.livemd (hunk 9 above)
Comment (alisinabh) [reply notebooks/text_generator.livemd]: This is great. I tested it and it seems the last part is not needed since we want to select everything anyways. 
So `x[[0..-1//1, -1]]` should do the trick.

File: notebooks/text_generator.livemd (hunk 12 above)
Comment (alisinabh) [reply notebooks/text_generator.livemd]: I created an `idx_to_char` mapping for explicitness. But AFAIK it wont help with efficiency as we only have `29` members and I believe BEAM treats maps smaller than `32` elements as lists. But it will make a difference if we expand the codepoints.

File: notebooks/text_generator.livemd (hunk 6 above)
Comment (alisinabh) [reply notebooks/text_generator.livemd]: Addressed in 7a502c4

File: notebooks/text_generator.livemd (hunk 12 above)
Comment (jonatanklosko) [reply notebooks/text_generator.livemd]: > I believe BEAM treats maps smaller than 32 elements as lists

Correct, but the keys are sorted, so lookup via binary search is O(log N) :)

File: notebooks/text_generator.livemd (hunk 9 above)
Comment (jonatanklosko) [reply notebooks/text_generator.livemd]: Yup, all remaining axes are sliced in full!

File: notebooks/text_generator.livemd [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+
+<!-- livebook:{"output":true} -->
+
+```
+:ok
+```
+
+## Generating Text
+
+After last step we will have our trained network parameters so we can start generating text with it. We simply need to give the initial sequence as the input to the network and select the most confident output. `Axon.predict/3` will give us the output layer and then using `Nx.argmax/1` we get the most confident neuron index then simply convert that index back to it's unicode representation.
```
Comment (jonatanklosko) [inline notebooks/text_generator.livemd:369]:
```suggestion
After last step we will have our trained network parameters so we can start generating text with it. We simply need to give the initial sequence as the input to the network and select the most confident output. `Axon.predict/3` will give us the output layer and then using `Nx.argmax/1` we get the most confident neuron index then simply convert that index back to it's Unicode representation.
```

**Review by jonatanklosko — APPROVED** [review]
:cat:

File: notebooks/text_generator.livemd [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+```
+
+Lowercase English characters in unicode encoding are presented with numbers between `97 = a` and `122 = z` a space is `32 = [ ]` a new line is `10 = \n` and the period is `46 = .`.
```
[the remark this answers is not in the document]
Comment (alisinabh) [reply notebooks/text_generator.livemd]: Addressed in feb8f98

File: notebooks/text_generator.livemd (hunk 14 above)
Comment (alisinabh) [reply notebooks/text_generator.livemd]: Addressed in feb8f98

**Comment by alisinabh** [comment]
Thanks @jonatanklosko 
Squashed and ready!
